### ISSUE TITLE
Crash in 0.36.0

### ISSUE THREAD

**Issue by RolandasRazma** [apollographql/apollo-ios#1473]
## Bug report

Crash using interceptors

<img width="1024" alt="Screenshot 2020-10-26 at 18 37 05" src="https://user-images.githubusercontent.com/1027187/97213977-4857c680-17ba-11eb-902d-90d617d3ef97.png">

## Versions

Please fill in the versions you're currently using: 

- `apollo-ios` SDK version: 0.36.0
- Swift version: 5.3

## Steps to reproduce

I have TokenAddingInterceptor that retrieves token from server and asynchronously calls `chain.proceedAsync`. I have check for TokenAddingInterceptor being released to not call up the chain on dealloc, so that's not the case.

**Comment by designatednerd** [comment]
The crash is happening in the `URLSessionClient` rather than in any of the interceptors - it looks like the client either a) Hasn't had a chance to set up its `URLSession` or b) has had the session torn down. 

Can you share how you're setting up your `TokenAddingService`?

**Comment by RolandasRazma** [comment]
There is nothing interesting to see there. It just check for token validity if it it expired updates it from server and calls back. I will try to trace/fix crash as it's blocking us, but as its another race condition its not going to be simple as I can't even replicate it reliably. Sometimes it crashes every run (tests) sometimes I hit https://github.com/apollographql/apollo-ios/issues/1376 and sometimes all tests passes without any problem :)

**Comment by RolandasRazma** [comment]
I think its same bug as https://github.com/apollographql/apollo-ios/issues/1376

I can replicate it with example from https://www.apollographql.com/docs/ios/tutorial/tutorial-mutations/

```swift
class TokenAddingInterceptor: ApolloInterceptor {
    func interceptAsync<Operation: GraphQLOperation>(chain: RequestChain, request: HTTPRequest<Operation>, response: HTTPResponse<Operation>?, completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard self != nil else { return }
            chain.proceedAsync(request: request, response: response, completion: completion)
        }
    }
    
}
```

**Comment by RolandasRazma** [comment]
This is how memory graph looks for crashing client

<img width="445" alt="Screenshot 2020-10-27 at 10 39 03" src="https://user-images.githubusercontent.com/1027187/97290611-b0e98680-1840-11eb-8bf9-e3690946ecf1.png">

**Comment by RolandasRazma** [comment]
and that makes me wonder...

NetworkFetchInterceptor:32 don't you need `self.currentTask?.cancel()` before it?

**Comment by RolandasRazma** [comment]
Ok, what's happening is that `RequestChain` (created in `RequestChainNetworkTransport`) lifetime is shorter than `LegacyInterceptorProvider` and `LegacyInterceptorProvider` does `shouldInvalidateClientOnDeinit` so what's happening is that `RequestChain` is created in `RequestChainNetworkTransport`, `LegacyInterceptorProvider` is released invalidating client and `RequestChain` knows nothing about it. 

I managed to "solve" it by retaining link to chain and canceling it in deist of RequestChainNetworkTransport` - not sure if that's correct thing to do, as that would imply that I need to cancel it on send as well (or retain all references created by send)

**Comment by RolandasRazma** [comment]
sorry for spam @designatednerd

**Comment by RolandasRazma** [comment]
ok, I have workaround to unblock us (to not retain chain)

```swift
class TokenAddingInterceptor: ApolloInterceptor {
    func interceptAsync<Operation: GraphQLOperation>(chain: RequestChain, request: HTTPRequest<Operation>, response: HTTPResponse<Operation>?, completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void) {
        DispatchQueue.main.async { [weak self, weak chain] in
            guard self != nil else { return }
            chain?.proceedAsync(request: request, response: response, completion: completion)
        }
    }
    
}
```

can't replicate with such change, but I see there is places where it is retained in apollo-ios itself (for example in LegacyCacheReadInterceptor), so most likely you still can hit it

**Comment by designatednerd** [comment]
right - if you let go of the chain there there's nothing hanging on to it, so ARC smashes it. 

I'm surprised the legacy interceptor provider is getting deallocated - that indicates that the Request Chain Network Transport itself is getting deallocated. I would really like to see how you're setting this stuff up for tests - it seems like maybe something is calling into an old instance of RCNT in different tests.

**Comment by RolandasRazma** [comment]
`RequestChainNetworkTransport` is deallocated, but the problem is not that it is deallocated, the problem is that it doesn't cleanup properly while it does that

```swift
class TokenAddingInterceptor: ApolloInterceptor {
    func interceptAsync<Operation: GraphQLOperation>(chain: RequestChain, request: HTTPRequest<Operation>, response: HTTPResponse<Operation>?, completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void) {
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) { [weak self] in
            guard self != nil else { return }
            chain?.proceedAsync(request: request, response: response, completion: completion)
        }
    }
}

let transport = RequestChainNetworkTransport(...), endpointURL: url)
let apolloClient = ApolloClient(networkTransport: transport, ...)
apolloClient.fetch(query: ...)
```

will make whole app crash in 1 second because `ApolloClient` will be deallocated together with `RequestChainNetworkTransport` while `RequestChain` will be retained by block (and it has to be)

**Comment by RolandasRazma** [comment]
this is the crash https://github.com/apollographql/apollo-ios/pull/1480

**Comment by RolandasRazma** [comment]
I guess easy workaround would be to remove force unwrap, but the real problem is that `RequestChain` don't know that it is no longer needed. Correct fix would be to add "cancel" for chain in `RequestChainNetworkTransport` as mentioned in https://github.com/apollographql/apollo-ios/issues/1473#issuecomment-717173144 the only difficulty it that there is multiple chains per transport and all of them needs to be cancelled

**Comment by RolandasRazma** [comment]
Created PR with workaround https://github.com/apollographql/apollo-ios/pull/1481 but somehow don't fee good about it :) It doe's work, but feels like some code smell

**Comment by RolandasRazma** [comment]
should I close this?

**Comment by designatednerd** [comment]
I'll close it when it actually ships

**Comment by designatednerd** [comment]
0.37.0 has shipped!
